### PR TITLE
Update GetLatestVersionAsync for NuGetPackageActions to include pre-release versions

### PR DESCRIPTION
**Pull request by jpinz**
Update GetLatestVersionAsync for NuGetPackageActions to include pre-release versions by default to match how the other package managers get the latest version. NPM and PyPI at least don't check for latest version not-including pre-releases.

We were having issues with getting the metadata for: https://www.nuget.org/packages/SlipeServer.Scripting/0.1.0-CI-20220607-083949 as all of the versions are pre-release versions.

### CODE REVIEW

**Review by pmalmsten — APPROVED** [review]

**Comment by gfs** [comment]
Hmm. This looks to me like the flag already existed so the bug is actually that the other managers are not respecting the flag in the default configuration - since they should not be returning prereleases with it set to false right?

This also seems to be to kind of be a semver change in that it modifies an existing public api. Is there a reason you can't for your cases pass true for the use prerelease flag instead of modifying the lib?

**Review by gfs — COMMENTED** [review]
See comment.

**Comment by jpinz** [comment]
> Hmm. This looks to me like the flag already existed so the bug is actually that the other managers are not respecting the flag in the default configuration - since they should not be returning prereleases with it set to false right?

So there is currently a disconnect between all project managers and the NuGet project manager, where the method I changed the default value of the flag is only implemented on the NuGet project manager. At the time it made sense to include that flag and default it to false, however, with all other project managers not having an equivalent flag, it made sense to change the default to match the functionality of the rest of the project managers.
 
> This also seems to be to kind of be a semver change in that it modifies an existing public api. Is there a reason you can't for your cases pass true for the use prerelease flag instead of modifying the lib?

The use prerelease flag isn't available that high up the call stack to the metadata call, but that could be added instead.
An alternative solution I had come up with is as follows:

``` cs
// If we don't want pre-releases but all versions are pre-releases, then we have to include pre-releases to get a latest version. 
if (!includePrerelease && versionsAscending.All(v => v.IsPrerelease))
{
    includePrerelease = true;
}
return versionsAscending
    .Last(v => includePrerelease || !v.IsPrerelease) // The latest version is the last in ascending order.
    .ToString();
```
Where essentially, if there are no versions of the package being called on that **aren't a pre-release version** then it will return the last pre-release version.

**Comment by pmalmsten** [comment]
One other thought I had on this - the package managers other than `NuGet` don't support `false` for this flag today, correct? As in, they don't change the result when the flag is `false`?

If that's true, should we remove the flag entirely? Or have the other package managers throw `NotSupportedException` when that flag is false?

**Comment by gfs** [comment]
> > Hmm. This looks to me like the flag already existed so the bug is actually that the other managers are not respecting the flag in the default configuration - since they should not be returning prereleases with it set to false right?
> 
> So there is currently a disconnect between all project managers and the NuGet project manager, where the method I changed the default value of the flag is only implemented on the NuGet project manager. At the time it made sense to include that flag and default it to false, however, with all other project managers not having an equivalent flag, it made sense to change the default to match the functionality of the rest of the project managers.

I think I agree with Paul. Changing the default value here doesn't fix the problem if no managers except NuGet respect this value - it just makes it harder to notice.

> 
> > This also seems to be to kind of be a semver change in that it modifies an existing public api. Is there a reason you can't for your cases pass true for the use prerelease flag instead of modifying the lib?
> 
> The use prerelease flag isn't available that high up the call stack to the metadata call, but that could be added instead. An alternative solution I had come up with is as follows:
> 
> ```cs
> // If we don't want pre-releases but all versions are pre-releases, then we have to include pre-releases to get a latest version. 
> if (!includePrerelease && versionsAscending.All(v => v.IsPrerelease))
> {
>     includePrerelease = true;
> }
> return versionsAscending
>     .Last(v => includePrerelease || !v.IsPrerelease) // The latest version is the last in ascending order.
>     .ToString();
> ```
> 
> Where essentially, if there are no versions of the package being called on that **aren't a pre-release version** then it will return the last pre-release version.

I'm not really a fan of this hidden behavior. If a user asks for the latest version but no pre-releases the result of the API call shouldn't be a pre-release - that violates the API contract where they said they didn't want pre-releases.

I don't think this is good to merge. I think either we need to implement the pre-release flag for every manager, or remove it. I don't think it makes sense to change the deafult.

**Comment by jpinz** [comment]
I'm cool with removing the pre-release flag for NuGet until we can implement it across all managers.

File: src/Shared/Contracts/IManagerPackageActions.cs [hunk 1]
```diff
@@ -38,25 +38,25 @@ public interface IManagerPackageActions<T> where T : IManagerPackageVersionMetad
...
+    /// Gets all versions of a package, both releases and pre-releases are included by default.
     /// </summary>
     /// <remarks>The first version in the list is the latest, and the oldest version is the last.</remarks>
     /// <param name="packageUrl">The <see cref="PackageURL"/> to get all available versions for.</param>
-    /// <param name="includePrerelease">If pre-release/beta versions should be included.</param>
+    /// <param name="includePrerelease">If pre-release/beta versions should be included, defaults to <c>true</c>.</param>
     /// <param name="useCache">If the cache should be checked for the available versions of this package.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to be used in the method call, defaults to <see cref="CancellationToken.None"/>.</param>
     /// <returns>An <see cref="IEnumerable{String}"/> of all of the versions for this package, in descending order.</returns>
     Task<IEnumerable<string>> GetAllVersionsAsync(PackageURL packageUrl, bool includePrerelease = true, bool useCache = true, CancellationToken cancellationToken = default);
```
Comment (gfs) [inline src/Shared/Contracts/IManagerPackageActions.cs:49]: If we are removing the flag it should be removed here too right?

File: src/Shared/Contracts/IManagerPackageActions.cs [hunk 2]
```diff
@@ -38,25 +38,25 @@ public interface IManagerPackageActions<T> where T : IManagerPackageVersionMetad
...
+    /// Gets the latest version of a package, both releases and pre-releases are included.
+    /// TODO: Reimplement include pre-release when all project managers support it.
     /// </summary>
     /// <param name="packageUrl">The <see cref="PackageURL"/> to get the latest version for.</param>
-    /// <param name="includePrerelease">If pre-release/beta versions should be included.</param>
     /// <param name="useCache">If the cache should be checked for the latest of this package.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to be used in the method call, defaults to <see cref="CancellationToken.None"/>.</param>
     /// <returns>The latest version of the <paramref name="packageUrl"/> or null if there are no versions.</returns>
-    Task<string?> GetLatestVersionAsync(PackageURL packageUrl, bool includePrerelease = false, bool useCache = true, CancellationToken cancellationToken = default);
+    Task<string?> GetLatestVersionAsync(PackageURL packageUrl, bool useCache = true, CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Contracts/IManagerPackageActions.cs:59]: Changing this to take an options object also has the advantage of future changes to add more arguments not breaking compatibility.

**Review by gfs — CHANGES_REQUESTED** [review]

File: src/Shared/Contracts/IManagerPackageActions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmalmsten) [reply src/Shared/Contracts/IManagerPackageActions.cs:59]: > Changing this to take an options object also has the advantage of future changes to add more arguments not breaking compatibility.

Yeah, that's an interesting point. It seems that [the convention in C# is that the CancellationToken always goes last](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1068)? That convention unfortunately gets in the way of adding new optional arguments to the end / rightmost position of a method signature, which would otherwise be fine.

File: src/Shared/Contracts/IManagerPackageActions.cs [hunk 3]
```diff
@@ -38,25 +38,25 @@ public interface IManagerPackageActions<T> where T : IManagerPackageVersionMetad
     Task<bool> DoesPackageExistAsync(PackageURL packageUrl, bool useCache = true, CancellationToken cancellationToken = default);
 
     /// <summary>
-    /// Gets all versions of a package.
+    /// Gets all versions of a package, both releases and pre-releases are included by default.
     /// </summary>
     /// <remarks>The first version in the list is the latest, and the oldest version is the last.</remarks>
     /// <param name="packageUrl">The <see cref="PackageURL"/> to get all available versions for.</param>
-    /// <param name="includePrerelease">If pre-release/beta versions should be included.</param>
+    /// <param name="includePrerelease">If pre-release/beta versions should be included, defaults to <c>true</c>.</param>
```
[the remark this answers is not in the document]
Comment (jpinz) [reply src/Shared/Contracts/IManagerPackageActions.cs:45]: I don't think so. `BaseProjectManager.EnumerateVersionsAsync` includes an optional `includePrerelease` flag that is defaulted to true. `TypedProjectManager` inherits `BaseProjectManager` and uses the `EnumerateVersionsAsync` implementation to call `GetAllVersionsAsync` with the `includePrerelease` flag as well. So all of the project managers have it, but again, the only one that appears to respect the value of it is NuGet.

[the remark this answers is not in the document]
Comment (gfs) [reply src/Shared/Contracts/IManagerPackageActions.cs:45]: > So all of the project managers have it, but again, the only one that appears to respect the value of it is NuGet.

I'm confused them by the different it seems we are back to exactly the same issue that caused us to remove it from the other API - it isn't being respected by the package managers.

[the remark this answers is not in the document]
Comment (jpinz) [reply src/Shared/Contracts/IManagerPackageActions.cs:45]: I can remove it everywhere else too, I don't think it matters that much. It would just be refactoring something outside of ManagerPackageActions.

[the remark this answers is not in the document]
Comment (jpinz) [reply src/Shared/Contracts/IManagerPackageActions.cs]: All prerelease flags are now removed.

**Review by gfs — APPROVED** [review]
Ideally this feature would exist in OSSGadget but it needs a better implementation - i think for example you should be able to specify to *only* enumerate pre-releases which is not possible under the current scheme, (and to be supported on all managers). See https://github.com/microsoft/OSSGadget/issues/368 to discuss adding it back in the future with a better implementation.

**Comment by gfs** [comment]
@jpinz one note, this is technically a semver change so we theoretically should update the version.json to 0.2 - however, ossgadget is also major version 0 which allows for breaking changes and this is minor so maybe its okay?

**Comment by jpinz** [comment]
> @jpinz as I think about this more, is this really needed? If removing the check works wouldn't passing "true" for pre-release on the existing API also fix the issue in the same way?

Yes it would, the issue is that for us it's failing as a part of getting metadata. Which doesn't expose the includePrerelease flag.

**Comment by gfs** [comment]
@jpinz which API call do you mean for get metadata? I am having trouble finding that in the changes.

**Review by gfs — CHANGES_REQUESTED** [review]
After further discussion we are pursuing adding the pre-release flag to the GetMetadataAsync api instead.

**Comment by jpinz** [comment]
> After further discussion we are pursuing adding the pre-release flag to the GetMetadataAsync api instead.

@gfs just updated the PR. Found a bug related to a lack of "tags" in nuget metadata so made a quick fix for that as well.
Added a lot of unit test coverage for the changes as well.

File: src/Shared/PackageActions/NuGetPackageActions.cs [hunk 4]
```diff
@@ -113,11 +114,17 @@ public class NuGetPackageActions : IManagerPackageActions<NuGetPackageVersionMet
     {
         FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
 
-        IEnumerable<NuGetVersion> versionsAscending = await resource.GetAllVersionsAsync(
+        ImmutableList<NuGetVersion> versionsAscending = (await resource.GetAllVersionsAsync(
             packageUrl.Name,
             _sourceCacheContext,
             _logger, 
-            cancellationToken);
+            cancellationToken)).ToImmutableList();
```
Comment (gfs) [inline src/Shared/PackageActions/NuGetPackageActions.cs:121]: Why this change to fully enumerate here?

File: src/Shared/PackageActions/NuGetPackageActions.cs [hunk 5]
```diff
@@ -134,12 +141,18 @@ public class NuGetPackageActions : IManagerPackageActions<NuGetPackageVersionMet
...
```
Comment (gfs) [inline src/Shared/PackageActions/NuGetPackageActions.cs:148]: Why this change to fully enumerate here?

File: src/oss-tests/Helpers/PackageActionsHelper.cs [hunk 6]
```diff
@@ -53,9 +53,19 @@ public static class PackageActionsHelper<T> where T : IManagerPackageVersionMeta
...
                     {
-                        return typeof(T) == typeof(NuGetPackageVersionMetadata)
-                            ? versionsArray.Where(v => includePrerelease || !NuGetVersion.Parse(v).IsPrerelease)
-                            : versionsArray;
+                        if (typeof(T) != typeof(NuGetPackageVersionMetadata))
+                        {
+                            return versionsArray;
+                        }
+
+                        if (!includePrerelease && versionsArray.All(v => NuGetVersion.Parse(v).IsPrerelease))
```
[the remark this answers is not in the document]
Comment (jpinz) [reply src/oss-tests/Helpers/PackageActionsHelper.cs:61]: This check isn't necessary you're right. Removed.

File: src/Shared/PackageActions/NuGetPackageActions.cs (hunk 4 above)
Comment (jpinz) [reply src/Shared/PackageActions/NuGetPackageActions.cs:121]: Removed.

File: src/Shared/PackageActions/NuGetPackageActions.cs [hunk 7]
```diff
@@ -113,11 +114,17 @@ public class NuGetPackageActions : IManagerPackageActions<NuGetPackageVersionMet
...
+        
+        // If we don't want pre-releases, but all versions are pre-releases, return an empty list.
+        if (!includePrerelease && versionsAscending.All(v => v.IsPrerelease))
```
[the remark this answers is not in the document]
Comment (jpinz) [reply src/Shared/PackageActions/NuGetPackageActions.cs:124]: Removed.

File: src/Shared/PackageActions/NuGetPackageActions.cs [hunk 8]
```diff
@@ -134,12 +141,18 @@ public class NuGetPackageActions : IManagerPackageActions<NuGetPackageVersionMet
...
-        IEnumerable<NuGetVersion> versionsAscending = await resource.GetAllVersionsAsync(
+        ImmutableList<NuGetVersion> versionsAscending = (await resource.GetAllVersionsAsync(
             packageUrl.Name,
             _sourceCacheContext,
             _logger, 
-            cancellationToken);
+            cancellationToken)).ToImmutableList();
 
+        // If we don't want pre-releases, but all versions are pre-releases, return null.
+        if (!includePrerelease && versionsAscending.All(v => v.IsPrerelease))
```
[the remark this answers is not in the document]
Comment (jpinz) [reply src/Shared/PackageActions/NuGetPackageActions.cs:151]: This check, unlike the others, is needed for `.Last` as that throws an exception if nothing matches the predicate.
However, I switched it to `.LastOrDefault<NuGetVersion?>` and now we don't need that condition anymore.

File: src/Shared/PackageActions/NuGetPackageActions.cs (hunk 5 above)
Comment (jpinz) [reply src/Shared/PackageActions/NuGetPackageActions.cs:148]: Removed.

**Review by gfs — APPROVED** [review]
Approved. I have the fix for the failing test in #371 which needs a quick review.
